File: examples/superluminal/realtime/realtime.py
import time
import numpy as np
import superluminal as lm

print("Welcome to Superluminal!")

data = np.random.rand(1, 8192).astype(np.complex64)

lm.plot(data, lm.line, label="Sine")

def callback():
    phase = 0

    while lm.running():
        # Generate sine wave
        for i in range(data.shape[1]):
            data[0, i] = np.sin(i * 0.01 + phase) * 20 + 30
        
        phase += 0.1

        # Update the plot.
        lm.update()

        # Sleep for 10 ms.
        time.sleep(0.01)

lm.realtime(callback)

print("Goodbye from Superluminal!")

File: examples/superluminal/realtime/realtime.cc
#include <jetstream/superluminal.hh>

using namespace Jetstream;

static Result App() {
    std::cout << "Welcome to Superluminal!" << std::endl;

    Tensor data(DeviceType::CPU, TypeToDataType<F32>(), {1, 8192});

    JST_CHECK(Superluminal::Plot("Sine", {{1}}, {
        .buffer = data,
        .type = Superluminal::Type::Line,
        .options = {},
    }));

    JST_CHECK(Superluminal::RealtimeLoop([&](const bool& running){
        F32 phase = 0.0f;

        while (running) {
            // Generate sine wave.

            for (U64 i = 0; i < data.shape(0); i++) {
                for (U64 j = 0; j < data.shape(1); j++) {
                    data.at<F32>(i, j) = std::sin(j * 0.01f + phase) * 0.25f;
                }
            }

            phase += 0.1f;

            Superluminal::Update();
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }
    }));

    std::cout << "Goodbye from Superluminal!" << std::endl;

    return Result::SUCCESS;
}

int main() {
    return App() == Result::SUCCESS ? 0 : 1;
}
